Userland/printf.cpp

```diff
@@ -0,0 +1,297 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/PrintfImplementation.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <AK/Types.h>
+#include <stdio.h>
+#include <unistd.h>
+
+[[gnu::noreturn]] void fail(const char* message)
+{
+    puts(message);
+    exit(1);
+}
+
+template<typename PutChFunc, typename ArgumentListT, template<typename T, typename U = ArgumentListT> typename NextArgument>
+struct PrintfImpl : public PrintfImplementation::PrintfImpl<PutChFunc, ArgumentListT, NextArgument> {
+    ALWAYS_INLINE int format_b(PutChFunc, char*&, int, const PrintfImplementation::State&, ArgumentListT&) const
+    {
+        fail("format specifier 'b' is not supported");
+    }
+    ALWAYS_INLINE int format_q(PutChFunc putch, char*& bufptr, int, const PrintfImplementation::State& state, ArgumentListT& ap) const
+    {
+        auto state_copy = state;
+        auto str = NextArgument<const char*>()(ap);
+        if (!str)
+            str = "(null)";
+
+        constexpr auto make_len_or_escape = [](auto str, bool mk_len, size_t field_width, auto putc) {
+            unsigned len = 2;
+            if (!mk_len)
+                putc('"');
+
+            for (size_t i = 0; str[i] && (mk_len ? true : (field_width >= len)); ++i) {
+                auto ch = str[i];
+                switch (ch) {
+                case '"':
+                case '$':
+                case '\\':
+                    ++len;
+                    if (!mk_len)
+                        putc('\\');
+                }
+                ++len;
+                if (!mk_len)
+                    putc(ch);
+            }
+
+            if (!mk_len)
+                putc('"');
+
+            return len;
+        };
+
+        auto len = make_len_or_escape(str, true, state_copy.field_width, [&](auto c) { putch(bufptr, c); });
+
+        if (!state_copy.dot && (!state_copy.field_width || state_copy.field_width < len))
+            state_copy.field_width = len;
+        size_t pad_amount = state_copy.field_width > len ? state_copy.field_width - len : 0;
+
+        if (!state_copy.left_pad) {
+            for (size_t i = 0; i < pad_amount; ++i)
+                putch(bufptr, ' ');
+        }
+
+        make_len_or_escape(str, false, state_copy.field_width, [&](auto c) { putch(bufptr, c); });
+
+        if (state_copy.left_pad) {
+            for (size_t i = 0; i < pad_amount; ++i)
+                putch(bufptr, ' ');
+        }
+
+        return state_copy.field_width;
+    }
+};
+
+template<typename T, typename V>
+struct ArgvNextArgument {
+    ALWAYS_INLINE T operator()(V) const
+    {
+        static_assert(sizeof(V) != sizeof(V), "Base instantiated");
+        return declval<T>();
```

Comment:
So *that's* what you need `declval` for?

Can't you just `ASSERT_NOT_REACHED()`?

Reply:
No, I need `declval` to be able to resolve `va_list&`.